src/storage/src/statistics.rs

```diff
@@ -372,6 +420,45 @@ impl SourceStatistics {
         }
     }
 
+    /// Clear the data, resetting gauges to uninitialized.
+    /// This does not reset prometheus gauges, which will be overwritten with new values.
+    pub fn clear(&self) {
+        let mut cur = self.stats.borrow_mut();
+        cur.0.clear();
+    }
+
+    /// Get a snapshot of the data, returning `None` if all gauges are not initialized.
+    pub fn snapshot(&self) -> Option<SourceStatisticsUpdate> {
+        let cur = self.stats.borrow_mut();
+
+        match &cur.0 {
+            SourceStatisticsRecord {
+                id,
+                worker_id,
+                messages_received,
+                bytes_received,
+                updates_staged,
+                updates_committed,
+                snapshot_committed: Some(snapshot_committed),
+                envelope_state_bytes: Some(envelope_state_bytes),
+                envelope_state_records: Some(envelope_state_records),
+                rehydration_latency_ms: Some(rehydration_latency_ms),
+            } => Some(SourceStatisticsUpdate {
+                id: *id,
+                worker_id: *worker_id,
+                messages_received: *messages_received,
+                bytes_received: *bytes_received,
+                updates_staged: *updates_staged,
+                updates_committed: *updates_committed,
+                snapshot_committed: *snapshot_committed,
+                envelope_state_bytes: *envelope_state_bytes,
+                envelope_state_records: *envelope_state_records,
+                rehydration_latency_ms: *rehydration_latency_ms,
+            }),
+            _ => None,
+        }
```

Comment:
Optional simplification:

```suggestion
            Some(SourceStatisticsUpdate {
                id: cur.0.id,
                worker_id: cur.0.worker_id,
                messages_received: cur.0.messages_received,
                bytes_received: cur.0.bytes_received,
                updates_staged: cur.0.updates_staged,
                updates_committed: cur.0.updates_committed,
                snapshot_committed: cur.0.snapshot_committed?,
                envelope_state_bytes: cur.0.envelope_state_bytes?,
                envelope_state_records: cur.0.envelope_state_records?,
                rehydration_latency_ms: cur.0.rehydration_latency_ms?,
            })
        }
```

Reply:
We only emit a `snapshot` if the gauges are initialized, so i cant directly take this, but this can be made cleaner

Reply:
oh i see the question marks now!